File: android/sample-ui/src/main/kotlin/com/usesmileid/sampleapps/ui/state/UseSmileIDSampleFlowResult.kt
package com.usesmileid.sampleapps.ui.state

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.saveable.Saver
import androidx.compose.runtime.saveable.listSaver
import androidx.compose.runtime.setValue
import com.usesmileid.sampleapps.ui.model.UseSmileIDSampleFlowRoute
import com.usesmileid.sampleapps.ui.model.UseSmileIDSampleFlowStatus
import com.usesmileid.sampleapps.ui.model.UseSmileIDSampleResult
import com.usesmileid.sampleapps.ui.model.UseSmileIDSampleScenario
import com.usesmileid.sampleapps.ui.model.UseSmileIDSampleThemeScenario

/** Saveable, because a count that resets on recreation would swallow the second call it exists to catch. */
class UseSmileIDSampleFlowResult(
    scenario: UseSmileIDSampleScenario = UseSmileIDSampleScenario.Normal,
    theme: UseSmileIDSampleThemeScenario = UseSmileIDSampleThemeScenario.BrandDefault,
    route: UseSmileIDSampleFlowRoute = UseSmileIDSampleFlowRoute.Fullscreen,
    status: UseSmileIDSampleFlowStatus = UseSmileIDSampleFlowStatus.Idle,
    jobId: String? = null,
    userId: String? = null,
    lastError: String? = null,
    resultCallbackCount: Int = 0,
    refreshCallbackCount: Int = 0,
) {
    var scenario by mutableStateOf(scenario)
        private set

    var theme by mutableStateOf(theme)
        private set

    var route by mutableStateOf(route)
        private set

    var status by mutableStateOf(status)
        private set

    var jobId by mutableStateOf(jobId)
        private set

    var userId by mutableStateOf(userId)
        private set

    var lastError by mutableStateOf(lastError)
        private set

    var resultCallbackCount by mutableIntStateOf(resultCallbackCount)
        private set

    var refreshCallbackCount by mutableIntStateOf(refreshCallbackCount)
        private set

    val snapshot: UseSmileIDSampleResult
        get() = UseSmileIDSampleResult(
            activeScenario = scenario,
            activeTheme = theme,
            route = route,
            jobStatus = status,
            resultCallbackCount = resultCallbackCount,
            refreshCallbackCount = refreshCallbackCount,
            jobId = jobId,
            userId = userId,
            lastError = lastError,
        )

    fun selectScenario(value: UseSmileIDSampleScenario) {
        scenario = value
    }

    fun selectTheme(value: UseSmileIDSampleThemeScenario) {
        theme = value
    }

    /** Navigation decides this, so it is set on arrival rather than on handoff. */
    fun enterRoute(value: UseSmileIDSampleFlowRoute) {
        route = value
    }

    /** Both counts reset here, so "exactly once" is asserted per run rather than per app launch. */
    fun startFlow() {
        status = UseSmileIDSampleFlowStatus.Running
        jobId = null
        userId = null
        lastError = null
        resultCallbackCount = 0
        refreshCallbackCount = 0
    }

    /** [userId] must be what the server returned; a local placeholder invalidates every run after it. */
    fun recordResultCallback(
        status: UseSmileIDSampleFlowStatus,
        jobId: String? = null,
        userId: String? = null,
        error: String? = null,
    ) {
        resultCallbackCount++
        this.status = status
        this.jobId = jobId
        this.userId = userId
        lastError = error
    }

    fun recordRefreshCallback() {
        refreshCallbackCount++
    }

    companion object {
        val Saver: Saver<UseSmileIDSampleFlowResult, Any> = listSaver(
            save = {
                listOf(
                    it.scenario.id,
                    it.theme.id,
                    it.route.id,
                    it.status.id,
                    it.jobId.orEmpty(),
                    it.userId.orEmpty(),
                    it.lastError.orEmpty(),
                    it.resultCallbackCount.toString(),
                    it.refreshCallbackCount.toString(),
                )
            },
            restore = { saved ->
                UseSmileIDSampleFlowResult(
                    // Read defensively: this runs after process death, where a throw takes the app down.
                    scenario = UseSmileIDSampleScenario.entries.firstOrNull { it.id == saved[0] }
                        ?: UseSmileIDSampleScenario.Normal,
                    theme = UseSmileIDSampleThemeScenario.entries.firstOrNull { it.id == saved[1] }
                        ?: UseSmileIDSampleThemeScenario.BrandDefault,
                    route = UseSmileIDSampleFlowRoute.entries.firstOrNull { it.id == saved[2] }
                        ?: UseSmileIDSampleFlowRoute.Fullscreen,
                    status = UseSmileIDSampleFlowStatus.entries.firstOrNull { it.id == saved[3] }
                        ?: UseSmileIDSampleFlowStatus.Idle,
                    jobId = saved[4].ifEmpty { null },
                    userId = saved[5].ifEmpty { null },
                    lastError = saved[6].ifEmpty { null },
                    resultCallbackCount = saved[7].toIntOrNull() ?: 0,
                    refreshCallbackCount = saved[8].toIntOrNull() ?: 0,
                )
            },
        )
    }
}

File: ios/SampleUI/Sources/SampleUI/State/UseSmileIDSampleFlowResult.swift
/// What the hosted flow reported. Held by the app state and nowhere else, so every launch is a fresh run and no count carries into the next.
public struct UseSmileIDSampleFlowResult: Equatable, Sendable {
  public private(set) var scenario: UseSmileIDSampleScenario
  public private(set) var theme: UseSmileIDSampleThemeScenario
  public private(set) var route: UseSmileIDSampleFlowRoute
  /// Recorded from the run's own snapshot, never re-read: sandbox until a run has said otherwise.
  public private(set) var environment: UseSmileIDSampleEnvironment
  public private(set) var status: UseSmileIDSampleFlowStatus
  public private(set) var jobId: String?
  public private(set) var userId: String?
  public private(set) var lastError: String?
  public private(set) var resultCallbackCount: Int
  public private(set) var refreshCallbackCount: Int

  public init(
    scenario: UseSmileIDSampleScenario = .normal,
    theme: UseSmileIDSampleThemeScenario = .brandDefault,
    route: UseSmileIDSampleFlowRoute = .fullscreen,
    environment: UseSmileIDSampleEnvironment = .sandbox,
    status: UseSmileIDSampleFlowStatus = .idle,
    jobId: String? = nil,
    userId: String? = nil,
    lastError: String? = nil,
    resultCallbackCount: Int = 0,
    refreshCallbackCount: Int = 0
  ) {
    self.scenario = scenario
    self.theme = theme
    self.route = route
    self.environment = environment
    self.status = status
    self.jobId = jobId
    self.userId = userId
    self.lastError = lastError
    self.resultCallbackCount = resultCallbackCount
    self.refreshCallbackCount = refreshCallbackCount
  }

  public var snapshot: UseSmileIDSampleResult {
    UseSmileIDSampleResult(
      activeScenario: scenario,
      activeTheme: theme,
      route: route,
      environment: environment,
      jobStatus: status,
      resultCallbackCount: resultCallbackCount,
      refreshCallbackCount: refreshCallbackCount,
      jobId: jobId,
      userId: userId,
      lastError: lastError
    )
  }

  public mutating func selectScenario(_ value: UseSmileIDSampleScenario) {
    scenario = value
  }

  public mutating func selectTheme(_ value: UseSmileIDSampleThemeScenario) {
    theme = value
  }

  public mutating func enterRoute(_ value: UseSmileIDSampleFlowRoute) {
    route = value
  }

  /// Both counts reset here, so "exactly once" is asserted per run rather than per app launch.
  public mutating func startFlow(environment: UseSmileIDSampleEnvironment) {
    self.environment = environment
    status = .running
    jobId = nil
    userId = nil
    lastError = nil
    resultCallbackCount = 0
    refreshCallbackCount = 0
  }

  /// `userId` must be what the server returned; a local placeholder invalidates every run after it.
  public mutating func recordResultCallback(
    status: UseSmileIDSampleFlowStatus,
    jobId: String? = nil,
    userId: String? = nil,
    error: String? = nil
  ) {
    resultCallbackCount += 1
    self.status = status
    self.jobId = jobId
    self.userId = userId
    lastError = error
  }

  /// The gate refused the run, so the SDK never mounted. Not counted as a result callback.
  public mutating func recordBlocked(reason: String, environment: UseSmileIDSampleEnvironment) {
    self.environment = environment
    status = .failed
    jobId = nil
    userId = nil
    lastError = reason
  }

  public mutating func recordRefreshCallback() {
    refreshCallbackCount += 1
  }

  /// The ten fields as strings in a fixed order, the Compose `Saver`'s shape.
  public var saved: [String] {
    [
      scenario.id, theme.id, route.id, environment.id, status.id,
      jobId ?? "", userId ?? "", lastError ?? "",
      String(resultCallbackCount), String(refreshCallbackCount)
    ]
  }

  /// Read defensively: this runs on a restore, where a throw takes the app down.
  public init(saved: [String]) {
    func field(_ index: Int) -> String {
      index < saved.count ? saved[index] : ""
    }
    self.init(
      scenario: UseSmileIDSampleScenario(id: field(0)),
      theme: UseSmileIDSampleThemeScenario(id: field(1)),
      route: UseSmileIDSampleFlowRoute(id: field(2)),
      environment: UseSmileIDSampleEnvironment(rawValue: field(3)) ?? .sandbox,
      status: UseSmileIDSampleFlowStatus(id: field(4)),
      jobId: field(5).isEmpty ? nil : field(5),
      userId: field(6).isEmpty ? nil : field(6),
      lastError: field(7).isEmpty ? nil : field(7),
      resultCallbackCount: Int(field(8)) ?? 0,
      refreshCallbackCount: Int(field(9)) ?? 0
    )
  }
}
